src/nnet3/attention.h

```diff
@@ -0,0 +1,329 @@
+//     u_t :=  Append(\sum_o c_{t,o} v_{t+o-num_left_inputs},  c_t)     [eqn:u]
+//
+//
+//  * Implementation
+//
+// The most time-consuming parts of this computation, we imagine, would be the
+// dot-products in [eqn:b] and the weighted sum (\sum_o) in [eqn:u].  They have
+// an awkward band-diagonal structure that would not be particularly convenient
+// to implement using CUBLAS and the like; I don't believe the relevant operations
+// exist in the BLAS interface, at least for [eqn:u].
+//
+// In the future I hope to implement this with block-wise matrix
+// multiplies-- imagine covering the band-diagonal part of a matrix with
+// rectangular blocks in such a way that all the nonzero elements are covered,
+// but the blocks might go over the zero parts a bit.   This could be done with
+// Or perhaps we can figure out how to implement the block-diagonal matrix
+// multiplies in CUDA.
+
+
+
+/**
+   This function is a utility function that is at the core of how we implement
+   attention.  It may in future need to be renamed and possibly moved into the
+   cudamatrix directory and implemented in CUDA.  The current implementation is
+   quite inefficient.  We can also consider doing a complete redesign of how the
+   implementation works, such that this function doesn't exist at all; or we
+   could have a batched version of this function that would operate on a batch
+   of A, B and C at once (or a "strided, batched" version where the difference
+   between the members of the batch is expressed as a stride).
+
+   This function implements a special operation that you could view as some kind
+   of matrix multiplication where only a band of the product is retained.
+
+   The inputs A and B must have the same number of columns
+   (A.NumCols() == B.NumCols()), and A and C must have the same
+   number of rows (A.NumRows() == C->NumRows()).  The number of
+   rows of B must exceed the number of rows of A.  Define
+      num_extra_rows = B.NumRows() - A.NumRows().
+   Then C.NumCols() - 1 must divide num_extra_rows.
+   Define
+      row_shift = num_extra_rows / (C.NumCols() - 1).
+
+   This function implements:
+      (*C)(i, j) = alpha * VecVec(A.Row(i), B.Row(i + j * row_shift))
+ */
+void GetAttentionDotProducts(Real alpha,
+                             const CuMatrixBase<BaseFloat> &A,
+                             const CuMatrixBase<BaseFloat> &B,
+                             CuMatrixBase<BaseFloat> *C);
+
+
+/**
+   This function is related to GetAttentionDotProducts(); it
+   is used in scaling the values by the softmax scales, and
+   in backprop.
+
+   We have put the A, B and C in an unusual order here in order
+   to make clearer the relationship with GetAttentionDotProducts().
+   The matrices have the same relationship in terms of their
+   dimensions, as A, B and C do in GetAttentionDotProducts().
+
+   This function implements:
+
+     A->Row(i) += alpha * C(i, j) * B.Row(i + j * row_shift).
```

Comment:
This line has `j` on the right-hand side but not on the left-hand side. I'm not sure I get it right.